File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/IntradayTickElementResponseError.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/IntradayTickElementResponseError.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class IntradayTickElementResponseError extends Element
{
    private final IntradayTickElementString _source, _category, _message, _subCategory;
    private final IntradayTickElementInt _code;

    IntradayTickElementResponseError()
    {
        int code = RandomDataGenerator.randomInt(99);
        String sourceGibberish = RandomDataGenerator.randomString(5).toLowerCase();

        this._source = new IntradayTickElementString("source", String.format("%s::%s%s", code, sourceGibberish, RandomDataGenerator.randomInt(99)));
        this._code = new IntradayTickElementInt("code", code);
        this._category = new IntradayTickElementString("category", "BAD_SEC");
        this._message = new IntradayTickElementString("message", String.format("Unknown/Invalid security [nid:%s]", code));
        this._subCategory = new IntradayTickElementString("subcategory", "INVALID_SECURITY");
    }
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("ErrorInfo"));
    }
    
    public Name name() throws Exception
    {
    	return new Name("responseError");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 5;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    public String getElementAsString(String name) throws Exception
    {
    	return this.getElement(name).getValueAsString();
    }
    
    public int getElementAsInt32(String name) throws Exception
    {
    	return this.getElement(name).getValueAsInt32();
    }
    
    public Element getElement(String name) throws Exception
    {
    	if(name.equals("source"))
    		return this._source;
    	
    	else if(name.equals("category"))
    		return this._category;
    	
    	else if(name.equals("message"))
    		return this._message;
    	
    	else if(name.equals("subcategory"))
    		return this._subCategory;
    	
    	else if(name.equals("code"))
    		return this._code;
    	
    	else
    		return super.getElement(name);
    }
    
    public boolean hasElement(String name)
    {
    	return name.equals("source") ||
    			name.equals("category") ||
    			name.equals("message") ||
    			name.equals("subcategory") ||
    			name.equals("code");
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();
        
        result.append(String.format("%s%s = {%s", tabs, this.name(), System.getProperty("line.separator")));

        result.append(this._source.prettyPrint(tabIndent + 1));
        result.append(this._code.prettyPrint(tabIndent + 1));
        result.append(this._category.prettyPrint(tabIndent + 1));
        result.append(this._message.prettyPrint(tabIndent + 1));
        result.append(this._subCategory.prettyPrint(tabIndent + 1));

        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));

        return result;
    }
}

File: BEmu_cpp/src/IntradayTickRequest/IntradayTickElementResponseError.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/IntradayTickRequest/IntradayTickElementResponseError.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include "IntradayTickRequest/IntradayTickElementResponseError.h"
#include "BloombergTypes/Name.h"
#include "BloombergTypes/ElementPtr.h"
#include "Types/RandomDataGenerator.h"
#include "IntradayTickRequest/IntradayTickElementString.h"
#include "IntradayTickRequest/IntradayTickElementInt.h"
#include "Types/IndentType.h"

namespace BEmu
{
	namespace IntradayTickRequest
	{
		IntradayTickElementResponseError::IntradayTickElementResponseError() :
			_category(new IntradayTickElementString("category", "BAD_SEC")),
			_subCategory(new IntradayTickElementString("subcategory", "INVALID_SECURITY"))

		{
			int code = RandomDataGenerator::RandomInt(99);
			std::string sourceGibberish = RandomDataGenerator::RandomString(5);
			std::transform(sourceGibberish.begin(), sourceGibberish.end(), sourceGibberish.begin(), ::tolower);

			//source
			std::stringstream sourceValue;
			sourceValue << code << "::" << sourceGibberish.substr(0, sourceGibberish.length() - 2) << RandomDataGenerator::RandomInt(99);
			this->_source = boost::shared_ptr<IntradayTickElementString>(new IntradayTickElementString("source", sourceValue.str()));

			//code
			this->_code = boost::shared_ptr<IntradayTickElementInt>(new IntradayTickElementInt("code", code));

			//message
			std::stringstream messageValue;
			messageValue << "Unknown/Invalid security [nid:" << code << "]";
			this->_message = boost::shared_ptr<IntradayTickElementString>(new IntradayTickElementString("message", messageValue.str()));
		}

		IntradayTickElementResponseError::~IntradayTickElementResponseError()
		{
		}

		Name IntradayTickElementResponseError::name() const
		{
			Name result("responseError");
			return result;
		}

		const char* IntradayTickElementResponseError::getElementAsString(const char* name) const
		{
			return this->getElement(name)->getValueAsString(0);
		}

		int IntradayTickElementResponseError::getElementAsInt32(const char* name) const
		{
			return this->getElement(name)->getValueAsInt32(0);
		}

		SchemaElementDefinition IntradayTickElementResponseError::elementDefinition() const
		{
			::blpapi_DataType_t dtype = (::blpapi_DataType_t)this->datatype();
			SchemaElementDefinition result(dtype, Name("ErrorInfo"));
			return result;
		}

		boost::shared_ptr<ElementPtr> IntradayTickElementResponseError::getElement(const char* name) const
		{
			if(strncmp("source", name, 7) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_source);

			else if(strncmp("code", name, 5) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_code);

			else if(strncmp("category", name, 9) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_category);

			else if(strncmp("message", name, 8) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_message);

			else if(strncmp("subcategory", name, 12) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_subCategory);

			else
				throw elementPtrEx;
		}

		bool IntradayTickElementResponseError::hasElement(const char* name, bool excludeNullElements) const
		{
			return
				(strncmp("source", name, 7) == 0) ||
				(strncmp("code", name, 5) == 0) ||
				(strncmp("category", name, 9) == 0) ||
				(strncmp("message", name, 8) == 0) ||
				(strncmp("subcategory", name, 12) == 0);
		}

		std::ostream& IntradayTickElementResponseError::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs = IndentType::Indent(level, spacesPerLevel);

			stream << tabs << "responseError = {" << std::endl;
			this->_source->print(stream, level + 1, spacesPerLevel);
			this->_code->print(stream, level + 1, spacesPerLevel);
			this->_category->print(stream, level + 1, spacesPerLevel);
			this->_message->print(stream, level + 1, spacesPerLevel);
			this->_subCategory->print(stream, level + 1, spacesPerLevel);
			stream << tabs << '}' << std::endl;

			return stream;
		}

	}
}